crates/bitcoin-da/src/helpers/parsers.rs

```diff
@@ -71,111 +84,109 @@ fn get_script(tx: &Transaction) -> Result<&Script, ParserError> {
-                        public_key.extend(bytes.as_bytes());
-                    } else if inside_envelope_index >= 9 {
-                        body.extend(bytes.as_bytes());
-                    }
-
-                    inside_envelope_index += 1;
-                } else if bytes.is_empty() {
-                    last_op = Some(OP_FALSE); // rust bitcoin pushes [] instead of op_false
-                }
-            }
-            Instruction::Op(another_op) => {
-                // don't allow anything except data pushes inside envelope
-                if inside_envelope {
-                    return Err(ParserError::EnvelopeHasNonPushOp);
-                }
-
-                last_op = Some(another_op);
-            }
+    // Parse header
+    let header_slice = read_push_bytes(instructions)?;
+    let Some(header) = TransactionHeader::from_bytes(header_slice.as_bytes()) else {
+        return Err(ParserError::InvalidHeaderLength);
+    };
+
+    // Check rollup name
+    if header.rollup_name != rollup_name.as_bytes() {
+        return Err(ParserError::InvalidRollupName);
+    }
+
+    // Parse transaction body according to type
+    match header.typ {
+        TransactionType::Inscribed => parse_type_0_body(instructions),
+        TransactionType::Unknown(_) => Err(ParserError::InvalidHeaderType),
+    }
+}
+
+fn read_instr<'a>(
+    instructions: &mut dyn Iterator<Item = Result<Instruction<'a>, ParserError>>,
+) -> Result<Instruction<'a>, ParserError> {
+    let instr = instructions
+        .next()
+        .unwrap_or(Err(ParserError::UnexpectedEndOfScript))?;
+    Ok(instr)
+}
+
+fn read_push_bytes<'a>(
+    instructions: &mut dyn Iterator<Item = Result<Instruction<'a>, ParserError>>,
+) -> Result<&'a StructPushBytes, ParserError> {
+    let instr = read_instr(instructions)?;
+    match instr {
+        PushBytes(push_bytes) => Ok(push_bytes),
+        _ => Err(ParserError::UnexpectedOpcode),
+    }
+}
+
+fn read_opcode(
+    instructions: &mut dyn Iterator<Item = Result<Instruction<'_>, ParserError>>,
+) -> Result<Opcode, ParserError> {
+    let instr = read_instr(instructions)?;
+    let Op(op) = instr else {
+        return Err(ParserError::UnexpectedOpcode);
+    };
+    Ok(op)
+}
+
+// Parse transaction body of Type0
+fn parse_type_0_body(
+    instructions: &mut dyn Iterator<Item = Result<Instruction<'_>, ParserError>>,
+) -> Result<ParsedInscription, ParserError> {
+    // PushBytes(XOnlyPublicKey)
+    let _public_key = read_push_bytes(instructions)?;
+    if OP_CHECKSIG != read_opcode(instructions)? {
+        return Err(ParserError::UnexpectedOpcode);
+    }
+
+    let op_false = read_push_bytes(instructions)?;
+    if !op_false.is_empty() {
+        return Err(ParserError::UnexpectedOpcode);
+    }
+
+    if OP_IF != read_opcode(instructions)? {
+        return Err(ParserError::UnexpectedOpcode);
+    }
+
+    let signature = read_push_bytes(instructions)?;
+    let public_key = read_push_bytes(instructions)?;
```

Comment:
how are you reading public key 2 times?

Reply:
The first one is xonlypublickey, the second one is sequencer/signer public key

Reply:
we are parsing the second script in the witness stack. it does not start with an xonly public key

Reply:
my bad, got confused